src/libs/tools/include/merging/mergingkdb.hpp

```diff
@@ -0,0 +1,105 @@
+/**
+ * @file
+ *
+ * @brief
+ *
+ * @copyright BSD License (see doc/COPYING or http://www.libelektra.org)
+ */
+
+#ifndef KDBMERGE_HPP_
+#define KDBMERGE_HPP_
+
+#include <kdb.hpp>
+#include <merging/threewaymerge.hpp>
+
+namespace kdb
+{
+
+namespace tools
+{
+
+namespace merging
+{
+
+/**
+ * Provides a merging wrapper around a KDB instance. The wrapper allows to pass
+ * a three way merger instance that is used to resolve conflicts during KDB set.
+ */
+class MergingKDB {
+public:
+	MergingKDB(KDB * kdb) : innerKdb(kdb) { }
+	~MergingKDB() { }
+
+	/**
+	 * Behaves like the KDB get function
+	 *
+	 * @see KDB
+	 */
+	virtual int get (KeySet & returned, std::string const & keyname);
+
+	/**
+	 * Behaves like the KDB get function
+	 *
+	 * @see KDB
+	 */
+	virtual int get (KeySet & returned, Key & parentKey);
+
+	/**
+	 * If a conflict occurs during set, the supplied merger is used to resolve the conflict.
+	 * If the conflict cannot be solved, an exception is thrown.
+	 *
+	 * @see KDB
+	 * @throws MergingKDBException
+	 */
+	virtual int set (KeySet & returned, std::string const & keyname, ThreeWayMerge & merger);
+
+	/**
+	 * If a conflict occurs during set, the supplied merger is used to resolve the conflict.
+	 * If the conflict cannot be solved, an exception is thrown. If the KeySet was successfully
+	 * written (either by merging or due the absence of a conflict) the supplied KeySet is updated
+	 * with the new content of the file.
+	 *
+	 * @see KDB
+	 * @throws MergingKDBException
+	 */
+	virtual int set (KeySet & returned, Key & parentKey, ThreeWayMerge & merger);
+
+private:
+	KDB * innerKdb;
+	KeySet base;
+};
+
+class MergingKDBException : public Exception
```

Comment:
> Also, should be derived from KDBException (or even use KDBException)?
> Should contain key, because maybe other warnings are present, too.

Agree.